src/TriggerBinding/SqlTriggerListener.cs

```diff
@@ -463,6 +470,124 @@ PRIMARY KEY ({primaryKeys})
+        {
+            return this.GetScaleStatusWithTelemetry(context.WorkerCount, context.Metrics?.ToArray());
+        }
+
+        private ScaleStatus GetScaleStatusWithTelemetry(int workerCount, SqlTriggerMetrics[] metrics)
+        {
+            var measures = new Dictionary<TelemetryMeasureName, double>
+            {
+                [TelemetryMeasureName.WorkerCount] = workerCount,
+                [TelemetryMeasureName.UnprocessedChangeCount] = (metrics is null || metrics.Length == 0) ? -1L : metrics[metrics.Length - 1].UnprocessedChangeCount,
+            };
+
+            TelemetryInstance.TrackEvent(TelemetryEventName.GetScaleStatusStart, this._telemetryProps, measures);
+
+            ScaleStatus scaleStatus = this.GetScaleStatusCore(workerCount, metrics);
+
+            measures.Add(TelemetryMeasureName.ScaleRecommendation, (double)scaleStatus.Vote);
+            TelemetryInstance.TrackEvent(TelemetryEventName.GetScaleStatusEnd, this._telemetryProps, measures);
+
+            return scaleStatus;
+        }
+
+        /// <summary>
+        /// Returns scale recommendation i.e. whether to scale in or out the host application. The recommendation is
+        /// made based on both the latest metrics and the trend of increase or decrease in the count of unprocessed
+        /// changes in the user table. In all of the calculations, it is attempted to keep the number of workers minimum
+        /// while also ensuring that the count of unprocessed changes per worker stays under the maximum limit.
+        /// </summary>
+        /// <param name="workerCount">The current worker count for the host application.</param>
+        /// <param name="metrics">The collection of metrics samples to make the scale decision.</param>
+        /// <returns></returns>
+        private ScaleStatus GetScaleStatusCore(int workerCount, SqlTriggerMetrics[] metrics)
+        {
+            const int minSamplesForScaling = 5;
+            const int maxChangesPerWorker = 1000;
+
+            var status = new ScaleStatus
+            {
+                Vote = ScaleVote.None
+            };
+
+            // Do not make a scale decision unless we have enough samples.
+            if (metrics is null || (metrics.Length < minSamplesForScaling))
+            {
+                this._logger.LogInformation($"Requesting no-scaling: Insufficient metrics for making scale decision for table: '{this._userTable.FullName}'.");
+                return status;
+            }
+
+            // Add worker if the count of unprocessed changes per worker exceeds the maximum limit.
+            long lastUnprocessedChangeCount = metrics.Last().UnprocessedChangeCount;
+            this._logger.LogInformation($"Unprocessed change count: {lastUnprocessedChangeCount}, worker count: {workerCount}, maximum changes per worker: {maxChangesPerWorker}.");
+
+            if (lastUnprocessedChangeCount > workerCount * maxChangesPerWorker)
```

Comment:
Do we really need this extra check? If it truly is increasing at a rate that requires scaling out then this will just be handled with the below checks. Right now if there's a sudden blip in the number of changes but then it immediately settles down do we really want to be scaling out and then immediately scaling back in?

Reply:
This is a valid concern, but I think this is a primary check and should be retained. If the load not strictly increasing [1100, 1050, 1150] but above the worker capacity, we should always ensure that per-worker load is maintained below their capacity.

The scale controller has complex logic for interval between scaling out. For example, it sometimes launches new instances at the same frequency of polling, which is 10 seconds while other times, especially for the very first scale-out request after a long period of no scaling (guessing based on my observation), it gives time for the newly scaled app to pick up and start processing the events and contribute to reduction in overall load before scaling out the app further.

Reply:
We can average across samples here to not react to one-off blips.

Reply:
I haven't made any change here. Will wait for your comment.

Reply:
Alright, I'm starting to lean towards not overthinking this and getting something out there and we can always change it later if needed. Given that the other triggers don't seem to have a problem with very basic logic for this stuff I think what we have is fine enough